runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WindmillStateReader.java

```diff
@@ -454,6 +469,10 @@ public void startBatchAndBlock() {
     // First, drain work out of the pending lookups into a set. These will be the items we fetch.
     HashSet<StateTag<?>> toFetch = Sets.newHashSet();
     try {
+      if (beforeRead != null) {
```

Comment:
Maybe we should move this down to only where we call getStateData
That will prevent a transition in cases where we don't actually issue  a read to the server.

Reply:
I was concerned that this might be incorrect if we ever call this from threads that are not the main processing thread (ie background pagination of bags).  However though we attempt to fetch the next bag page ahead of time it isn't issued in the background, just whenever it is needed or another state fetch would block.

Reply:
Done (after next force push)